File: 347-top-k-frequent-elements/top-k-frequent-elements.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var topKFrequent = function (nums, k) {
    let countMap = new Map()

    for (const num of nums) {
        countMap.set(num, (countMap.get(num) || 0) + 1);
    }

    const maxHeap = [...countMap.entries()].sort((a, b) => b[1] - a[1]);
    
    const result = [];
    for (let i = 0; i < k; i++) {
        result.push(maxHeap[i][0]);
    }

    return result;
};

File: 347-top-k-frequent-elements/top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        # Step 1: Create a frequency map (dictionary in Python)
        count_map = {}
        
        # Step 2: Count frequency of each number
        for num in nums:
            count_map[num] = count_map.get(num, 0) + 1
        
        # Step 3: Sort the dictionary items by frequency (descending)
        max_heap = sorted(count_map.items(), key=lambda x: x[1], reverse=True)
        
        # Step 4: Extract the top k elements
        result = []
        for i in range(k):
            result.append(max_heap[i][0])
        
        return result
        